### PR TITLE
fix #580, match cases, resolve types, involving generics

### LINKED ISSUE #580: match cases involving generics: require-condition1 failed,
```
test is
  abort(T type, a T) is

  a(T type, b T | abort T) is
    match b
      T => say "t"
      abort => say "i"

  a 2
  a i32 (abort 2)

  unit

```
```
java.lang.Error: require-condition1 failed: dev.flang.ast.Type:featureOfType:771
        at dev.flang.util.ANY.require(ANY.java:78)
        at dev.flang.ast.Type.featureOfType(Type.java:771)
        at dev.flang.ast.Case.resolveType(Case.java:306)
        at dev.flang.ast.Case.resolveType(Case.java:238)
        at dev.flang.ast.Match.resolveTypes(Match.java:185)
        at dev.flang.ast.Feature$ResolveTypes.action(Feature.java:1288)
        at dev.flang.ast.Match.visit(Match.java:132)
        at dev.flang.ast.Match.visit(Match.java:48)
        at dev.flang.ast.Block.visit(Block.java:227)
        at dev.flang.ast.Block.visit(Block.java:41)
        at dev.flang.ast.Impl.visit(Impl.java:312)
        at dev.flang.ast.Feature.visit(Feature.java:1038)
        at dev.flang.ast.Feature.internalResolveTypes(Feature.java:1319)
        at dev.flang.ast.Resolution.resolveTypes(Resolution.java:461)
        at dev.flang.ast.AbstractFeature.resultTypeIfPresent(AbstractFeature.java:650)
        at dev.flang.ast.Call.resolveTypes(Call.java:1719)
        at dev.flang.ast.Feature$ResolveTypes.action(Feature.java:1282)
        at dev.flang.ast.Feature$ResolveTypes.action(Feature.java:1274)
        at dev.flang.ast.Call.visit(Call.java:1056)
        at dev.flang.ast.Call.visit(Call.java:49)
        at dev.flang.ast.Block.visit(Block.java:227)
        at dev.flang.ast.Block.visit(Block.java:41)
        at dev.flang.ast.Impl.visit(Impl.java:312)
        at dev.flang.ast.Feature.visit(Feature.java:1038)
        at dev.flang.ast.Feature.internalResolveTypes(Feature.java:1319)
        at dev.flang.ast.Resolution.resolveOne(Resolution.java:371)
        at dev.flang.ast.Resolution.resolve(Resolution.java:338)
        at dev.flang.fe.SourceModule.createASTandResolve(SourceModule.java:260)
        at dev.flang.fe.FrontEnd.<init>(FrontEnd.java:175)
        at dev.flang.tools.Fuzion.lambda$parseArgsForBackend$3(Fuzion.java:825)
        at dev.flang.tools.Tool.run(Tool.java:153)
        at dev.flang.tools.Fuzion.main(Fuzion.java:476)
```

### PR DESCRIPTION
**Pull request by michaellilltokiwa**

### CODE REVIEW

**Comment by michaellilltokiwa** [comment]
this fix is needed for PR #445

File: tests/reg_issue580_match_cases_involvin_generics/Makefile [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#  Tokiwa Software GmbH, Germany
+#
+#  Source code of Fuzion test Makefile
+#
+#  Author: Fridtjof Siebert (siebert@tokiwa.software)
+#
+# -----------------------------------------------------------------------
+
+override NAME = issue580
+include ../positive.mk
```
Comment (fridis) [inline tests/reg_issue580_match_cases_involvin_generics/Makefile:27]: would maybe be better to use `simple.mk` and add the expected output. 
`positive.mk`-tests usually contain code to print `PASS`/`FAIL` and return with an exit code `1` on failure.

File: src/dev/flang/ast/Case.java [hunk 2]
```diff
@@ -303,7 +303,7 @@ AbstractType resolveType(Resolution res, AbstractType t, List<AbstractType> cgs,
         if (CHECKS) check
           (Errors.count() > 0 || cg != null);
         if (cg != null &&
-            (inferGenerics  && t.featureOfType() == cg.featureOfType() /* match feature, take generics from cg */ ||
+            (inferGenerics  && !cg.isGenericArgument() && t.featureOfType() == cg.featureOfType() /* match feature, take generics from cg */ ||
```
Comment (fridis) [inline src/dev/flang/ast/Case.java:306]: does this produce the right error if you duplicate the case-line that matches `T` in your example?

Comment (michaellilltokiwa) [reply src/dev/flang/ast/Case.java:306]: yes:
```
--CURDIR--/tests/reg_issue580_match_cases_involvin_generics/issue580.fz:33:7: error 1: 'case' clause matches type that had been matched already.
      T => say "t"
------^
Case matches type 'issue580.a.T'.
Originally matched at --CURDIR--/tests/reg_issue580_match_cases_involvin_generics/issue580.fz:32:7:
      T => say "t"
------^.
Subject type is one of 'issue580.a.T' or 'issue580.abort<issue580.a.T>'.

one error.

```

**Pull request merged**